New far forward beampipe and cryostat model v2025 implementation

nat93 opened the pull request:
### Briefly, what does this PR introduce?
This PR introduces a complete model of the cryostat (fwd and bwd) with detailed description of superconducting coils, support tubes, yokes, yoke shielding, heat shielding, and cryostat vessel. The fwd and rear sides were adjusted in order to avoid overlaps with off-momentum detectors and low-Q^2 tagger pipes.

The corresponding STEP file can be found on SharePoint: EIC Public Sharing Docs → Documents → Experimental Program → ePIC → Engineering → STR-Files → IR6_CRYOSTAT_2200m_top_4-24-2025.stp [https://brookhavenlab.sharepoint.com/:u:/s/EICPublicSharingDocs/EdR44ODny1BEmMTTks61CCwBe8tJYMc0iWkPIYY_EjYEuw?e=2XLUEt]

The changes have been communicated to TIC, IR, and MDP communities.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?

### Does this PR change default behavior?
Yes, it adds extra materials between the beam pipe (beam loss hot-spots) and the ePIC detector. It is crucial for the realistic detector background and dose calculations.

wdconinc commented on compact/definitions.xml:
```diff
@@ -319,6 +319,15 @@ The unused IDs below are saved for future use.
     <constant name="TaggerTracker_ID"        value="198"/>
     <constant name="TaggerCalorimeter_ID"    value="199"/>
 
+    <documentation>
+      #### (200-210) Far Backward Beamline Magnets
+    </documentation>
+    <constant name="Q1APR_ID"  value="200"/>
+    <constant name="Q1BPR_ID"  value="201"/>
+    <constant name="Q2PR_ID"   value="202"/>
+    <constant name="Q1ER_ID"   value="203"/>
+    <constant name="Q2ER_ID"   value="204"/>
+
```
We don't allocate IDs to passive elements like magnets.

nat93 replied:
Here I simply followed the previous implementation. Similar ID were already assigned to other magnets, so why should I avoid doing this here? Should we then remove ID for all magnets?

wdconinc commented on compact/far_backward/beamline_extension_electron.xml:
```diff
@@ -3,11 +3,53 @@
+    <detector id="Q1ER_ID" name="Q1ER_BeamlineMagnet" type="ip6_CylindricalDipoleMagnet">
+      <coil name="Q1ER_Quad_Coil" vis="MagnetCoilVis" shape="ConeSegment">
+        <placement x="Q1ER_Center_posX" y="Q1ER_Center_posY" z="Q1ER_Center_posZ" theta="Q1ER_Rotation_Angle" phiNum="3" phiStart="90 * deg" phiStep="90 * deg"/>
+        <dimensions rmin1="6.748780 * cm" rmax1="7.040880 * cm" rmin2="5.938520 * cm" rmax2="6.230620 * cm" dz="180.0 * cm / 2." sphi="-60.0 * deg / 2." dphi="60.0 * deg"/>
```
These look like calculated numbers. I assume they aren't specified to nm precision. Please include reference to where these numbers are from.

nat93 replied:
I took the number from the STEP file provided by the magnet devision. I believe initially they designed the cryostat in inches. Is there any specific difficulties using the numbers I put?

wdconinc reviewed (commented):
Please contribute gradually instead of through 1.5k lines per pull request. There is no realistic way to review large change sets like this for correctness, and certainly not on the time scale that we then are then pressured to include this. Please split this PR into more easily reviewable contributions.

Please include references to where the quantities that are taken from drawings or other design materials are able to be validated, typically as comments in the xml.

wdconinc commented:
> This specific modification has undergone extensive discussion and received approval from the TIC, IR, and MDP committees over the past few weeks or even months.

Whatever was approved before is the direction of what should be implemented: yes, an updated beamline must be implemented. I don't at all disagree with that.

Here, we're discussing at a more technical level: if we (ePIC collaboration) are asked in a year, potentially when you are not at BNL anymore or not working on EIC, to modify this geometry or verify that it is indeed correct, do we have sufficient information and in sufficiently clear form for someone (else) to do this? That is what this review process is about.

For decadal efforts, we have to take a longer term view here than just implementing it in any way that seems to work in the moment.

> Please advise on the best location to upload the STEP file for the cryostat so it can be accessed easily?

Since the Frascati meeting it was agreed by the EIC project that we now have to move beyond the sharing of individual detector parameters on https://eic.jlab.org/Geometry/ and that the project will start sharing CAD documents in STEP format with the ePIC collaboration. This is in the hands of the project now. I would suggest talking to Elke about how to get the 2025 reference model added there. (FYI, @johnlajoie, you know the latest on this)

nat93 commented:
> > This specific modification has undergone extensive discussion and received approval from the TIC, IR, and MDP committees over the past few weeks or even months.
> 
> Whatever was approved before is the direction of what should be implemented: yes, an updated beamline must be implemented. I don't at all disagree with that.
> 
> Here, we're discussing at a more technical level: if we (ePIC collaboration) are asked in a year, potentially when you are not at BNL anymore or not working on EIC, to modify this geometry or verify that it is indeed correct, do we have sufficient information and in sufficiently clear form for someone (else) to do this? That is what this review process is about.
> 
> For decadal efforts, we have to take a longer term view here than just implementing it in any way that seems to work in the moment.
> 
> > Please advise on the best location to upload the STEP file for the cryostat so it can be accessed easily?
> 
> Since the Frascati meeting it was agreed by the EIC project that we now have to move beyond the sharing of individual detector parameters on https://eic.jlab.org/Geometry/ and that the project will start sharing CAD documents in STEP format with the ePIC collaboration. This is in the hands of the project now. I would suggest talking to Elke about how to get the 2025 reference model added there. (FYI, @johnlajoie, you know the latest on this)

Dear @wdconinc ,
Thanks.

To correct the cryostat geometry in the future (if needed), one would need to modify the sizes of the relevant elements in the XML file based on the STEP file. Most of the elements have their dimensions defined independently, not linked to other components. I tried to at least link the position and orientation to the yokes.

I can also add more explanatory comments in the code to clarify where the numerical values come from and how they can be adjusted, so future modifications to the XML files are easier to follow.

I will ask Elke to add the STEP file to the geometry database, and once that is done, I will include the link in this PR and in the code.

wdconinc reviewed (dismissed):
Looks good to me (even if large).

simonge commented:
There are some significant issues with this at the moment from my side. This is coming out of the order in which changes in the far backward region need to be addressed. Narrowing of the beampipe with the current geometry shows the beamspot grazing the exit of B2eR resulting in an almost complete loss of acceptance in the Low-Q2 Taggers. (BeamspotID4 in the following plot.)

![image](https://github.com/user-attachments/assets/f597f9c0-5a28-4fa4-a8d9-803f31a5b84f)

I can try and better explain my proposed approach later on.

ajentsch reviewed (commented):
The changes for the FF detector geometries are fine for this PR. The B0 folks are working on updating the B0 tracker to better reflect reality, and I will need to re-tune the OMD positions and associated matrices in a separate PR once this is merged. Thanks for all of this work, Andrii.

wdconinc commented:
> Hi @wdconinc , there are 5 checks failed after some recent updates, mainly associated with eicweb and eic-shell. 
> Should we ignore them? Thanks.

We don't ignore tests. They get fixed elsewhere and then we will merge here. It's being worked on.

simonge commented:
> Thank you for checking this. But why we should keep using the old outdated geometry saying that something is not like before when we know that B2eR must be outside the cryostat (this will be addressed in next PRs) and the beam pipe must be much narrower than it is now in the DD4Hep (the vacuum group is working on its design)?

The outdated geometry has been kept in place as there are a number of developments I've been working on which really should be in place in order to properly understand the effects of any change while not breaking continuity for the people doing analysis from the simulation.

- Beamline benchmark: Added to show the steering of any new configuration matches that expected by the accelerator group.
- Acceptance benchmark: Needed to see the potential acceptance of the taggers based on the B2eR opening. https://github.com/eic/detector_benchmarks/pull/167
- Reconstruction training benchmark: Along with updating the magnetic optics an update to the parameters used to reconstruct the original electron momentum need to be updated or the physics analysis won't work. This allows this to be reconstructed based on the output position and direction from B2eR. https://github.com/eic/detector_benchmarks/pull/171
- Train this on the 6.3 branch and verify its performance is appropriate.
- Add this new trained reconstruction to https://github.com/eic/epic-data/
- Update the geometry with the parameters which are needed to change to 6.3 without making and physical changes. https://github.com/eic/epic/pull/890
- Update to the new(er) 6.3 lattice also linking the new neural network for reconstruction in the calibrations. https://github.com/eic/epic/pull/838
- At this point the changes from this PR should be trivial and have limited effect on any performance.

This all is with the view to making future updates to the beamline much easier.

The fact that the vacuum group is working on the design based on the much narrower beampipe without incorporating any requirement of the Low-Q2 tagger performance is quite concerning.

> Hi @wdconinc , there are 5 checks failed after some recent updates, mainly associated with eicweb and eic-shell.
Should we ignore them? Thanks.

Based on the plot I shared before, the newly added beamline benchmark will fail for the PR as it is (or at least should). It requires that no electrons are lost in the beamline optics. Although that's not why they're failing at the moment.

I would suggest if the background studies cannot wait that you rebase this to https://github.com/eic/epic/tree/6.3-Lattice which should already set up with the correct magnet positions and shapes and carry out the studies using the independent branch while waiting.

I can also again suggest changes which will break this PR into smaller steps which are more manageable.

nat93 commented:
@wdconinc and @simonge ,
I have restored the outgoing electron beam pipe sizes and adjusted the rear cryostat openings to avoid overlaps. Now there should not be any changes to the previous far-bwd geometry in terms of magnetic field (location, strength), beam pipes, and vacuum. So, I expect no change in the low-Q^2 acceptance. Only the outside the beam pipe changed (cryostat).
Please let me know your thoughts. Thanks.

wdconinc commented:
I'll wait until @simonge can approve this after the revert of the far backward changes.

nat93 commented:
Hi @simonge , please let me know your thoughts about the latest update: no changes to far-bwd beam pipe. If you are okay with that - please approve the PR and we will merge it.

simonge commented:
Thanks for your updates Andrii.

I'm still not entirety comfortable with this PR. There is too much to review with due attention to ensure it is both appropriately maintainable and functional without dedicating a significant uninterrupted period of time. It would take less time to review and merge a series of incremental PRs. This might just be a flaw in my current attention span capacity.

Some questions, for my understanding, and comments which I believe would improve the code.

Why have you moved the far_backward magnet definitions out of magnets and into beamline_extension_electrons? This won't be included in the default simulation geometry, this might be your intention to keep it simpler? The extensions were primarily there to roughly illustrate the space constraints for the luminosity detectors. Any material between the central detector and far backward detectors might change what we see in the simulations, will likely not be too important but probably should be included by default.

> This cpp file is used only for cryostat magnets, so there is no need to create a separate - it is already a separate class. For most of the elements, the common coordinate is the yoke position. However, there are many individual components that are not aligned with the yoke, such as holes, electron beam pipe elements and magnets.

This will likely highlight my ignorance of the magnet geometry. The Q1eR and Q2eR coils and support tubes are already factorized out from the cryostat volume with the cryostat volume built around them from the Q1APR etc definition. Could we not do the same for the Q1APR etc. making the xml files and geometry hierarchy easier to understand? It looks like the code can just be reused as you have nicely designed it to treat all of the components independently where the material is just defined by its name. Of course the magnet itself in the experiment won't work without the cryostat but it makes the files easier to interpret and maintain.

> I will need to re-tune the OMD positions and associated matrices in a separate PR once this is merged.

I don't like this approach of intentionally breaking a detector without at least having something lined up to immediately fix it. I understand this will provide the means of producing more realistic background and does calculations for the central detector but will in the meantime break holistic reconstruction attempts. It might not be a lot of people looking at this but from experience analysers can end up wasting significant time hunting down why their plots suddenly don't look right before asking the people who know. The solution here may be to define the OMD and RP reconstruction matrix as part of the detector description along with the fields so that if any changes are made to the positions/fields the matrix can be updated at the same time.

A couple of other general comments.

Please remove the white space between the numbers, units and operators in the xml. Some of the parameters would be easier to find, maintain and understand if they were defined as constants in the xml files.

Please update the copyright statements to include your name.

ajentsch commented:
> 
> > I will need to re-tune the OMD positions and associated matrices in a separate PR once this is merged.
> 
> I don't like this approach of intentionally breaking a detector without at least having something lined up to immediately fix it. I understand this will provide the means of producing more realistic background and does calculations for the central detector but will in the meantime break holistic reconstruction attempts. It might not be a lot of people looking at this but from experience analysers can end up wasting significant time hunting down why their plots suddenly don't look right before asking the people who know. The solution here may be to define the OMD and RP reconstruction matrix as part of the detector description along with the fields so that if any changes are made to the positions/fields the matrix can be updated at the same time.

@simonge While I appreciate the concern, since I am one of the people running the simulation campaign now, I will of course ensure that the necessary changes for the OMD and RP are in place before a new campaign would be launched with the changes - and really, it's only the OMD detectors, which at the moment are only being used by me for full reconstruction of final-state particles. 

I have to disagree with using the detector definitions to define the RP and OMD matrices, as this would further complicate the definitions used for the detector - having them in EICrecon is the more sensible option to make it easier for me to maintain since they are used directly for momentum reconstruction, and will ultimately be replaced by the ML + polynmomal reconstruction methods, when time exists.

wdconinc commented:
> I will of course ensure that the necessary changes for the OMD and RP are in place before a new campaign would be launched with the changes

Considering how much we keep having to delay campaigns for late changes and requests, I would suggest that this work of implementing any changes to the OMD reconstruction matrices should start soon. It seems unlikely that we would merge this and knowingly break (apparently) also the far forward without any PR already in place.

nat93 commented:
Hi @simonge,

> I'm still not entirety comfortable with this PR. There is too much to review with due attention to ensure it is both appropriately maintainable and functional without dedicating a significant uninterrupted period of time. It would take less time to review and merge a series of incremental PRs. This might just be a flaw in my current attention span capacity.

There is only one .cpp file with two functions and four XML files describing the cryostat geometry. Nothing has changed for any of the systems except for the OMD, for which @ajentsch has provided a comprehensive explanation.
I believe it would be wise to focus our efforts on ensuring the geometry is correct, rather than on optimizing the design or detector performance - especially given that the current setup in DD4Hep is outdated and will never be built. Furthermore, we are all aware that additional updates are on the way, which will inevitably affect the detectors: a new beam pipe from the vacuum group, new magnets, and more.
We are already behind with the B2eR update, which, as decided by the Project, must be moved outside the cryostat. Of course, we can propose further updates to the warm B2eR, but only after implementing the changes that have already been decided.

Furthermore, radiation calculations are an ultra-high priority. They are essential for machine and detector protection, including the design and placement of collimators, shielding, optics tuning, and the vacuum system. I do not see how any auxiliary detector studies could be more important than this. We must implement the geometry and run full radiation simulations - including TID, background rates, and related metrics - to determine whether ePIC can withstand the expected conditions.

> Why have you moved the far_backward magnet definitions out of magnets and into beamline_extension_electrons? 

I did this for consistency. The HSR BWD/FWD magnets are defined in beamline_extension_hadron.xml/ion_beamline.xml, while the ESR BWD/FWD magnets are defined in beamline_extension_electron.xml/electron_beamline.xml. This separation makes it clear where to look for each ring.

> > This cpp file is used only for cryostat magnets, so there is no need to create a separate - it is already a separate class. For most of the elements, the common coordinate is the yoke position. However, there are many individual components that are not aligned with the yoke, such as holes, electron beam pipe elements and magnets.
> This will likely highlight my ignorance of the magnet geometry. The Q1eR and Q2eR coils and support tubes are already factorized out from the cryostat volume with the cryostat volume built around them from the Q1APR etc definition. Could we not do the same for the Q1APR etc. making the xml files and geometry hierarchy easier to understand?

In the latest commit, I have moved BWD cryo magnets from far_backward/beamline_extension_e/h.xml to far_backward/magnets.xml.

> Please remove the white space between the numbers, units and operators in the xml. 

I have removed the white space between the numbers, units and operators in the xml. 

> Please update the copyright statements to include your name.

I have added my name.

nat93 commented:
> @nat93 Everyone's work is important in this collaboration. Acting dismissive of others' work will not be tolerated here.

Hi @wdconinc , perhaps my message was not clear. The main idea is to optimize the procedure to avoid double work for everyone. I did not mean what you wrote. Opposite, I do respect other people work and it would be not efficient optimizing configuration that is already outdated, postponing the central detector protection and collider design - this is the message.

nat93 commented on src/forwardBeamPipeBrazil.cpp:
```diff
@@ -517,10 +525,23 @@ static Ref_t create_detector(Detector& det, xml_h e, SensitiveDetector /* sens *
 
   SubtractionSolid final_vacuum_main_pipe(
       vacuum_main_pipe, cutout_for_OMD_station,
-      Position(0.0, 0.0, (2251.0 - beampipe_dimensions[pieceIdx].zCenter)));
```
Hi @ajentsch , should not we use units here as well?

nat93 commented:
Hi @ajentsch,
Thanks for catching the missing units — you are absolutely right that all dimensions should be explicitly defined with units. I have also added a few other missing units for OMDs and corrected some typos in the older scripts, as we previously discussed ('apperture' -> 'aperture').

ajentsch commented:
Yes, we should. Those were put in "by hand" to have vacuum surrounding the
detector planes. Those numbers will need to updated now, too. Will try to
do that tomorrow.

-------------------------------------------------------------
Alex Jentsch
Assistant Staff Scientist, EIC Directorate
Brookhaven National Laboratory
Upton, NY 11973

Bldg. 510, 2-197
Phone (office): 631-344-2139
Phone (cell): 281-726-0114

Pronouns: he/him/his

On Fri, Jul 4, 2025, 12:20 AM Andrii Natochii ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/forwardBeamPipeBrazil.cpp
> <https://github.com/eic/epic/pull/893#discussion_r2184301338>:
>
> > @@ -517,10 +525,23 @@ static Ref_t create_detector(Detector& det, xml_h e, SensitiveDetector /* sens *
>
>    SubtractionSolid final_vacuum_main_pipe(
>        vacuum_main_pipe, cutout_for_OMD_station,
> -      Position(0.0, 0.0, (2251.0 - beampipe_dimensions[pieceIdx].zCenter)));
>
> Hi @ajentsch <https://github.com/ajentsch> , should not we use units here
> as well?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eic/epic/pull/893#pullrequestreview-2985423223>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADABSF3MWPU3VGW5ZS6XHH33GX6JPAVCNFSM6AAAAAB7UPHSWCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSOBVGQZDGMRSGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

johnlajoie commented:
Hey @nat93 @simonge @ajentsch @wdconinc  - I'm trying to follow along with this an other PR's to see where we are in converging on an updated geometry that can be used in a consistent fashion for the "events with background" that we need to get going on the preTDR studies. @simonge - you had a bulleted list of seven things you wanted to see before this PR was merged. Is this feasible in a reasonable amount of time, and how much time is that? 

I realize a lot of work is being done by a small number of people so please don't take any of this as a criticism, just trying to to make sure I understand where we are.

ajentsch commented:
EICrecon PR with updated OMD matrices is ready: https://github.com/eic/EICrecon/pull/1957

nat93 commented:
Dear @wdconinc, @simonge, and @johnlajoie,

Following the suggestions from Simon and Wouter, as well as our Zoom discussion, @ajentsch and I have refined the PR to meet the outlined criteria. - Many thanks to Alex for assistance.

The FWD beam pipe remains unchanged in this PR - there is no impact on the far-forward detectors. On the BWD side, the beam pipe was adjusted to eliminate overlaps between the cryostat and the OMD chamber (Alex has updated the detector positions accordingly). The magnetic fields remain unchanged, and the detector acceptance is not reduced. In fact, on the FWD side, the acceptance may slightly increase due to the larger inner radius of the HSR cryostat compared to the beam stay-clear aperture.

This PR is a step forward, but additional improvements will follow: proper beam pipe definitions, a warm and split B2eR, and refinements to the magnetic fields and/or coil length (to be coordinated with the optics and magnet teams).

If everyone agrees, we propose to proceed with merging this PR.

veprbl commented:
Trying to figure out what are the remaining blockers.
@simonge You were planning to look at the impact of this latest version?
We also need to check material map before/after merging this.

simonge commented:
From what I see there is no impact compared to the current version at the moment as expected.

simonge commented:
> We also need to check material map before/after merging this.
There shouldn't be any change to the material map as I understand because everything is outside of the central detector region and outside the beampipe in the FF/FB regions.

There might be some changes I suggest down the line but happy to approve now and get the OMD update in EICrecon merged too.

simonge reviewed (approved):
Approved. I was hoping someone with more experience of the material map would confirm a new one wasn't needed.

veprbl commented:
I will take care of the map.
